File: algorithms/code/leetcode/lc349_intersection_of_two_arrays/LC349IntersectionOfTwoArrays.java
package leetcode.lc349_intersection_of_two_arrays;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

/**
 * https://leetcode.com/problems/intersection-of-two-arrays/ #easy
 */
public final class LC349IntersectionOfTwoArrays {
    public int[] intersection(final int[] nums1, final int[] nums2) {
        Arrays.sort(nums1);
        Arrays.sort(nums2);
        Set<Integer> found = new HashSet<>();
        for (int i1 = 0, i2 = 0; i1 < nums1.length && i2 < nums2.length;) {
            if (nums1[i1] < nums2[i2]) {
                i1++;
                continue;
            }
            if (nums1[i1] > nums2[i2]) {
                i2++;
                continue;
            }
            found.add(nums1[i1]);
            i1++;
            i2++;
        }
        int[] result = new int[found.size()];
        int index = 0;
        for (int value : found) {
            result[index] = value;
            index++;
        }
        return result;
    }
}
package leetcode.lc349_intersection_of_two_arrays;

import org.junit.jupiter.api.Test;

import java.util.Arrays;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;

public final class LC349IntersectionOfTwoArraysTests {
    @Test
    public void testExample() throws Exception {
        int[] nums1 = { 1, 2, 2, 1 };
        int[] nums2 = { 2, 2 };
        int[] result = new LC349IntersectionOfTwoArrays().intersection(nums1, nums2);
        Arrays.sort(result);
        int[] expected = { 2 };
        assertArrayEquals(expected, result);
    }

    @Test
    public void testExampleFlipped() throws Exception {
        int[] nums1 = { 2, 2 };
        int[] nums2 = { 1, 2, 2, 1 };
        int[] result = new LC349IntersectionOfTwoArrays().intersection(nums1, nums2);
        Arrays.sort(result);
        int[] expected = { 2 };
        assertArrayEquals(expected, result);
    }

    @Test
    public void test12And11() throws Exception {
        int[] expected = { 1 };
        int[] result = new LC349IntersectionOfTwoArrays().intersection(new int[] { 1, 2 }, new int[] { 1, 1 });
        Arrays.sort(result);
        assertArrayEquals(expected, result);
    }

    @Test
    public void test495And94984() throws Exception {
        int[] expected = { 4, 5, 9 };
        int[] result = new LC349IntersectionOfTwoArrays().intersection(new int[] { 4, 9, 5 },
                new int[] { 9, 4, 9, 8, 5 });
        Arrays.sort(result);
        assertArrayEquals(expected, result);
    }
}

File: algorithms/code/leetcode/lc349_intersection_of_two_arrays/lc349_intersection_of_two_arrays.rs
// https://leetcode.com/problems/intersection-of-two-arrays/
// #easy

use std::collections::HashSet;

pub struct Solution;

impl Solution {
    pub fn intersection(nums1: &mut [i32], nums2: &mut [i32]) -> Vec<i32> {
        nums1.sort_unstable();
        nums2.sort_unstable();
        let mut found = HashSet::new();
        let mut i1 = 0;
        let mut i2 = 0;
        while i1 < nums1.len() && i2 < nums2.len() {
            if nums1[i1] < nums2[i2] {
                i1 += 1;
                continue;
            }
            if nums1[i1] > nums2[i2] {
                i2 += 1;
                continue;
            }
            found.insert(nums1[i1]);
            i1 += 1;
            i2 += 1;
        }
        let mut result = Vec::new();
        for value in found {
            result.push(value);
        }
        result
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_example() {
        let mut nums1 = vec![1, 2, 2, 1];
        let mut nums2 = vec![2, 2];
        let mut result = Solution::intersection(&mut nums1, &mut nums2);
        result.sort_unstable();
        let expected = vec![2];
        assert_eq!(result, expected);
    }

    #[test]
    fn test_example_flipped() {
        let mut nums1 = vec![2, 2];
        let mut nums2 = vec![1, 2, 2, 1];
        let mut result = Solution::intersection(&mut nums1, &mut nums2);
        result.sort_unstable();
        assert_eq!(result, vec![2]);
    }

    #[test]
    fn test_12_and_11() {
        let mut nums1 = vec![1, 2];
        let mut nums2 = vec![1, 1];
        let mut result = Solution::intersection(&mut nums1, &mut nums2);
        result.sort_unstable();
        let expected = vec![1];
        assert_eq!(result, expected);
    }

    #[test]
    fn test_495_and_94984() {
        let mut nums1 = vec![4, 9, 5];
        let mut nums2 = vec![9, 4, 9, 8, 5];
        let mut result = Solution::intersection(&mut nums1, &mut nums2);
        result.sort_unstable();
        let expected = vec![4, 5, 9];
        assert_eq!(result, expected);
    }
}
